README.md
# BasicCRUDApp

# Motivation
Coding a basic web app implementing CRUD principles. It also
implements deletion with comment and undeletion features.


# Overview
CRUD interface for items and shipments of items, parts of which supports
deletion with comment and restore. Frontend is written in React with classes
and backend is Fastify with mysql2 as database interface. Testing is achieved
with Mocha, Chai and Sinon. Database migration is done with a node script
utilizing postgrator. Plain SQL migration code is found in migrations.


# Changelog
Any notes about changes can be found in CHANGELOG.md


# Requirements
- EITHER: docker, docker-compose
- OR: node.js with npm, mysql/mariadb


# Setup

## Docker
- Copy/rename .env.example to .env
- Set a (random) DB_PASS in .env
- run: docker-compose up


## Manual
- Copy/rename .env.example to .env
- Set all fields in .env to an existing empty database
- run: npm ci
- run: npm run build:prod
- run: npm run migrate
- run: npm run start:prod


# Package scripts

## lint
Lints all files with eslint.

## build:dev
Builds the project in development mode.

## build:prod
Builds the project in production mode.

## start:dev
Starts the project in development mode. Will build the project beforehand.
Any changes in frontend code will cause a webpack rebuild.

## start:prod
Starts the project in production mode. Will NOT build anything, so you would have
to call build:prod beforehand.

## migrate <version>
Migrates the database specified in .env to the most recent version, or <version>
if specified.

## migrate:test <version>
Same thing as migrate, but for the test database whose parameters are prefixed with TEST_
in .env.

## migrate-start
First migrates the production database to the most recent version and then starts the project.
This script is used by the docker version.

## test
Builds and tests the project with its own testsuite. It is required that the TEST_ settings
in .env point to a valid database which has been migrated to the current version with migrate-test.

## test-only
Does not build the project, but just runs the testsuite.


# Database Summary
The app uses 6 different database tables. Here is a short doc about them.

## deletions
Contains deletion comments for objects that have been marked for deletion but have not yet been removed.

### Schema
- id BIGINT PRIMARY KEY NOT NULL AUTO_INCREMENT - the primary id of this deletion
- created_at TIMESTAMP NOT NULL NOW() - the time this entry was created
- updated_at  TIMESTAMP NOT NULL NOW() ON UPDATE NOW() - the last time this entry was last updated
- comment VARCHAR(255) NOT NULL - the deletion comment


## items
Contains the central information for an item.

### Schema
- id BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY - the primary id of this item
- deletion_id BIGINT - the id of an entry in deletions if this item is marked for deletion
- created_at TIMESTAMP NOT NULL DEFAULT NOW() - the time this entry was created
- updated_at TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW() - the time this entry was last updated
- name VARCHAR(64) NOT NULL UNIQUE - the name of this item
- FOREIGN KEY (deletion_id) REFERENCES deletions (id) ON DELETE SET NULL


## external_item_assignments
Contains info to any item assignments that are not shipments.

### Schema
- id BIGINT NOT NULL AUTO_INCREMENT,
- PRIMARY KEY (id)


## item_assignments
Contains assignments of items. Basically changes in the number of items.

### Schema
- id BIGINT NOT NULL AUTO_INCREMENT - the primary id of this item_assignment
- item_id BIGINT NOT NULL - the id of the item that is assigned
- shipment_id BIGINT - the id of the shipment if the item is assigned to a shipment, otherwise NULL
- external_assignment_id BIGINT - the id of the external_item_assignment if it is assigned, otherwise NULL
- created_at TIMESTAMP NOT NULL DEFAULT NOW() - the time this entry was created
- updated_at TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW() - the time this entry was last updated
- assigned_count INT NOT NULL - the amount of items that are assigned
- PRIMARY KEY (id),
- FOREIGN KEY (item_id) REFERENCES items (id),
- FOREIGN KEY (shipment_id) REFERENCES shipments (id) ON DELETE CASCADE,
- FOREIGN KEY (external_assignment_id) REFERENCES external_item_assignments (id) ON DELETE CASCADE,
- UNIQUE KEY (item_id, shipment_id, external_assignment_id),
- CONSTRAINT exactly_one_assignment_type_chk CHECK ((CASE WHEN shipment_id IS NULL THEN 0 ELSE 1 END + CASE WHEN external_assignment_id IS NULL THEN 0 ELSE 1 END) = 1)

### Triggers
- item_assignments_insert AFTER INSERT - throws an SQLEXCEPTION if the total amount of the item in question is negative
- item_assignments_insert AFTER UPDATE - throws an SQLEXCEPTION if the total amount of the item in question is negative
- item_assignments_insert AFTER DELETE - throws an SQLEXCEPTION if the total amount of the item in question is negative

### Associated Functions
- AVAIL_ITEMS_COUNT(item_id BIGINT) - returns the amount of available items


## shipments
Contains shipments of items, collections that are meant to be sent out.

### Schema
- id BIGINT NOT NULL AUTO_INCREMENT - the primary id of this shipment
- created_at TIMESTAMP NOT NULL DEFAULT NOW() - the time this entry was created
- updated_at TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW() - the time this entry was last updated
- name VARCHAR(64) NOT NULL - the name of this shipment
- destination VARCHAR(64) NOT NULl - the destination of this shipment
- PRIMARY KEY (id)


## schemaversion
Table used by postgrator to store migration information


CHANGELOG.md
# CHANGELOG

## 0.1.0
### March 5, 2022
First release, every older development version is completely incompatible with this one.


src/index.ts
import express from "express"
import bodyParser from "body-parser"
import mysql2, { OkPacket, QueryError, RowDataPacket } from "mysql2"
import dotenv from "dotenv"
import path from "path"
import winston from "winston"


enum Error {
    DB = "Database Error",
    FIELD = "Field Error"
}

type ErrorResponse = {
    name: Error,
    errno?: number,
    message?: string
}

declare global {
    namespace Express {
         export interface Request {
                custom: {
                    inventoryEntryId?: number
                }
         }
    }
}


function logDbError(error: QueryError, hostname?: string) {
    if (hostname) {
        logger.error(`${hostname} caused database error ${error.code}: ${error.message}`)
    } else {
        logger.error(`Database error ${error.code}: ${error.message}`)
    }
}


function isInteger(n: string): boolean {
    return /^-?\d+$/.test(n)
}

dotenv.config()

const logger = winston.createLogger({
    level: process.env.LOG_LEVEL,
    format: winston.format.json(),
    transports: [
        new winston.transports.File({ filename: 'error.log', level: 'error' }),
        new winston.transports.File({ filename: 'combined.log' }),
    ]
})


function isMysqlError(error: ErrorResponse | QueryError): error is QueryError {
    return Boolean((error as QueryError).name)
}

function handleMixedError(error: ErrorResponse | QueryError,
    req: express.Request, res: express.Response) {
    if (isMysqlError(error)) {
        logDbError(error, req.hostname)

        const body: ErrorResponse = { name: Error.DB }
        res.status(500).send(body)
    } else {
        res.send(error)
    }

    return 0
}


const app = express()
app.use(bodyParser.json())

app.use((req, res, next) => {
    req.custom = {}
    next()
})

app.use("/inventory/item/:id", (req, res, next) => {
    if (isInteger(req.params.id)) {
        req.custom.inventoryEntryId = Number.parseInt(req.params.id, 10)
        next()
    } else {
        logger.info(`${req.hostname} tried to access inventory without a valid`
        + `entry id (${req.custom.inventoryEntryId})`)

        const body: ErrorResponse = {
            name: Error.FIELD,
            message: "The id has to be specified as number in the url path"
        }

        res.status(400).send()
    }
})

app.get("/inventory", (req, res) => {
    const stmt = "SELECT * FROM inventory WHERE deletion_id IS NULL"

    dbPromise.query(stmt)
    .then(([results, fields]) => {
        logger.info(`${req.hostname} requested all entries`)

        res.send(results)
    }, (error) => {
        handleMixedError(error, req, res)
    })
})

app.get("/inventory/deleted", (req, res) => {
    const stmt = "SELECT inventory.id, inventory.name, inventory.count, "
    + "deletions.comment FROM inventory INNER JOIN deletions "
    + "ON inventory.deletion_id=deletions.id WHERE deletion_id IS NOT NULL"

    dbPromise.query(stmt)
    .then(([results, fields]) => {
        logger.info(`${req.hostname} requested all deleted entries`)

        res.send(results)
    }, (error) => {
        handleMixedError(error, req, res)
    })
})

app.get("/inventory/item/:id", (req, res) => {
    const stmt = "SELECT * FROM inventory WHERE id = ?"

    dbPromise.query(stmt, req.custom.inventoryEntryId)
    .then(([results, fields]) => {
        results = results as RowDataPacket[]
        logger.info(`${req.hostname} requested entry with id ${req.custom.inventoryEntryId}`)

        res.send(results[0])
    }, (error) => {
        handleMixedError(error, req, res)
    })
})

app.put("/inventory", (req, res) => {
    if (req.body.hasOwnProperty("name") && req.body.hasOwnProperty("count")
        && req.body.name.length !== 0 && isInteger(req.body.count) && req.body.count >= 0) {
        const stmt = "INSERT INTO inventory SET ?"

        dbPromise.query(stmt, req.body)
        .then(([results, fields]) => {
            results = results as OkPacket
            logger.info(`${req.hostname} created entry with id ${results.insertId} in inventory`)

            res.status(201).send({
                name: req.body.name,
                count: req.body.count,
                id: results.insertId
            })
        }, (error) => {
            return handleMixedError(error, req, res)
        })
    } else {
        logger.info(`Received malformed/missing creation parameters from ${req.hostname}`)

        const body: ErrorResponse = {
            name: Error.FIELD,
            message: "Missing or malformed json fields"
        }

        res.status(400).send(body)
    }
})

app.put("/inventory/item/:id", (req, res) => {
    const update = req.body.hasOwnProperty("name") || req.body.hasOwnProperty("count")

    if (update) {
        const stmt = `UPDATE inventory SET ? WHERE id = ${mysql2.escape(req.custom.inventoryEntryId)}`

        dbPromise.query(stmt, req.body)
        .then(([results, fields]) => {
            results = results as OkPacket

            if (results.affectedRows > 0) {
                logger.info(`${req.hostname} updated entry with id ${req.custom.inventoryEntryId}`)
            } else {
                logger.info(`${req.hostname} tried to update non-existent entry `
                + `with id ${req.custom.inventoryEntryId} from inventory`)
            }

            return Promise.resolve(0)
        })
        .then(() => {
            res.send()
        }, (error) => {
            return handleMixedError(error, req, res)
        })
     } else {
        Promise.resolve()
        .then(() => {
            const stmt = "SELECT deletion_id FROM inventory "
            + "WHERE id = ? AND deletion_id IS NOT NULL"

            return dbPromise.query(stmt, req.custom.inventoryEntryId)
        })
        .then(([results, fields]) => {
            results = results as RowDataPacket[]

            if (results.length === 1) {
                const deletionId = results[0].deletion_id
                const stmt = "DELETE FROM deletions WHERE id = ?"

                return dbPromise.query(stmt, deletionId)
            } else {
                const body: ErrorResponse = {
                    name: Error.FIELD,
                    message: "The entry with the specified id is not in "
                    + "the deleted entries"
                }

                return Promise.reject(body)
            }
        })
        .then(() => {
            res.send()
        }, (error) => {
            return handleMixedError(error, req, res)
        })
    }
})

app.delete("/inventory/item/:id", (req, res) => {
    new Promise((resolve, reject) => {
        if (req.body.hasOwnProperty("comment")) {
            resolve(req.body.comment)
        } else {
            logger.info(`${req.hostname} requested to delete entry in `
            + `inventory without a deletion comment`)

            const body: ErrorResponse = {
                name: Error.FIELD,
                message: "There has to be a deletion comment for a deletion"
            }

            res.status(400)
            reject(body)
        }
    })
    .then((comment) => {
        const stmt = "INSERT INTO deletions (comment) VALUES (?)"
        return dbPromise.query(stmt, comment)
    })
    .then(([results, fields]) => {
        logger.info(`${req.hostname} added a deletion comment for entry with `
        + `id ${req.custom.inventoryEntryId} in inventory`)

        results = results as OkPacket
        const deletionId = mysql2.escape(results.insertId)

        const stmt = `UPDATE inventory SET ? WHERE id = ${mysql2.escape(req.custom.inventoryEntryId)}`

        return dbPromise.query(stmt, { deletion_id: deletionId })
    })
    .then(() => {
        logger.info(`${req.hostname} marked entry with id ${req.custom.inventoryEntryId} in `
        + `inventory as deleted`)

        res.send()
    }, (error) => {
        return handleMixedError(error, req, res)
    })
})

app.use(express.static(path.resolve(__dirname, "..", "public")))


const dbConnConfig: mysql2.ConnectionOptions = {
    host: process.env.DB_HOST,
    port: Number.parseInt(process.env.DB_PORT, 10),
    user: process.env.DB_USER,
    password: process.env.DB_PASS,
    database: process.env.DB_NAME
}

const dbConnection = mysql2.createConnection(dbConnConfig)

dbConnection.connect(err => {
    if (!err) {
        logger.info(`Connected to database ${dbConnConfig.database} `
        + `on ${dbConnConfig.host}:${dbConnConfig.port} `
        + `as user ${dbConnConfig.user}`)
    } else {
        logDbError(err)
    }
})

const dbPromise = dbConnection.promise()
dbPromise.query(
    "CREATE TABLE IF NOT EXISTS deletions "
    + "(id BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY, "
    + "created_at TIMESTAMP NOT NULL DEFAULT NOW(), "
    + "update_at TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW(), "
    + "comment VARCHAR(255) NOT NULL)")
    .then(() => {
        return dbPromise.query("CREATE TABLE IF NOT EXISTS inventory "
        + "(id BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY, "
        + "deletion_id BIGINT, "
        + "created_at TIMESTAMP NOT NULL DEFAULT NOW(), "
        + "updated_at TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW(), "
        + "name VARCHAR(64) NOT NULL, "
        + "count INT NOT NULL DEFAULT 0, "
        + "FOREIGN KEY (deletion_id) REFERENCES deletions (id) ON DELETE SET NULL)")
    })
    .then(() => {
        return new Promise((resolve, reject) => {
            const port = process.env.PORT
            app.listen(port, () => {
                logger.info(`Started web server listening on port ${port}`)
            })

            resolve(0)
        })
    })
    .catch(logDbError)
